### PR TITLE
Support safe MySQL upgrades

### PR DESCRIPTION
**Pull request by frouioui**
This Pull Request adds code logic that will set the `global.innodb_fast_shutdown` MySQL variable to `0` if there is a need for a safe MySQL upgrade.

Part of https://github.com/vitessio/vitess/issues/14160

### CODE REVIEW

File: pkg/controller/vitessshardreplication/reconcile_drain_test.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+			current:   "docker.io/vitess/mysql:8.0.23",
+			desired:   "docker.io/vitess/mysql:8.0.23",
+			needsSafe: false,
+		},
+		{
+			name:      "equal version with different registry",
+			current:   "docker.io/vitess/mysql:8.0.23",
+			desired:   "docker.io/vitess/mysql:8.0.23",
+			needsSafe: false,
+		},
```
Comment (GuptaManan100) [inline pkg/controller/vitessshardreplication/reconcile_drain_test.go:58]: They are the same cases

**Review by GuptaManan100 — APPROVED** [review]
Rest LGTM!

File: pkg/controller/vitessshardreplication/reconcile_drain_test.go (hunk 1 above)
Comment (frouioui) [reply pkg/controller/vitessshardreplication/reconcile_drain_test.go]: Fixed via c30e759

File: pkg/controller/vitessshardreplication/reconcile_drain.go [hunk 2]
```diff
@@ -255,8 +261,16 @@ func (r *ReconcileVitessShard) reconcileDrain(ctx context.Context, vts *planetsc
 		}
 	}
 
+	// 4. Check if we need to any operations like disabling fast shutdown
```
Comment (deepthi) [inline pkg/controller/vitessshardreplication/reconcile_drain.go:264]:
```suggestion
	// 4. Check if we need to perform any operations like disabling fast shutdown
```

File: pkg/controller/vitessshardreplication/reconcile_drain.go [hunk 3]
```diff
@@ -458,6 +472,151 @@ func candidatePrimary(ctx context.Context, wr *wrangler.Wrangler, shard *topo.Sh
...
+	if dstParts[0] == curParts[1] && dstParts[1] < curParts[1] {
+		return false, fmt.Errorf("cannot downgrade minor version from %s to %s", current, desired)
+	}
+
+	// Alright, here it gets more tricky. MySQL has had a complicated release history. For the 8.0 series,
+	// up to 8.0.34 at least (known at this point), it was not supported to downgrade patch releases
+	// as patch release could also include on-disk data format changes. This happened a number of times
+	// in practice as well, so this concern is real.
+	//
+	// MySQL though as announced a new release strategy, see:
```
Comment (deepthi) [inline pkg/controller/vitessshardreplication/reconcile_drain.go:594]:
```suggestion
	// MySQL though has announced a new release strategy, see:
```

**Review by deepthi — APPROVED** [review]
Nice work! I only found a couple of typos in comments. Please fix them and then merge.

**Pull request merged**